spacy/tokens/doc.pyx

```diff
@@ -1451,6 +1451,140 @@ cdef class Doc:
+        }
+
+        # Gather token-level properties.
+        for token in doc_json["tokens"]:
+            words.append(doc_json["text"][token["start"]:token["end"]])
+            for annotation_type in token_annotations:
+                # Assuming all tokens have exactly the same set of attributes.
+                if token_annotations[annotation_type]["req"] in token:
+                    if token_annotations[annotation_type]["values"] is None:
+                        token_annotations[annotation_type]["values"] = []
+                    token_annotations[annotation_type]["values"].append(token[annotation_type])
+
+        # Initialize doc instance.
+        start = 0
+        cdef const LexemeC* lex
+        cdef str orth_
+        cdef bint has_space
+        reconstructed_words, spaces = get_words_and_spaces(words, doc_json["text"])
+        assert words == reconstructed_words
+
+        for i, token in enumerate(doc_json["tokens"]):
+            has_space = spaces[i]
+            orth_ = doc_json["text"][token["start"]:token["end"]]
+            lex = self.vocab.get(self.mem, orth_)
+            self.push_back(lex, has_space)
+
+        # Set remaining token-level attributes via Doc.from_array().
+        if token_annotations["head"]["values"] is not None:
+            token_annotations["head"]["values"] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations["head"]["values"])
+            ]
+        if token_annotations["dep"]["values"] is not None:
+            MISSING_DEP_ = self.vocab.strings[MISSING_DEP]
+            token_annotations["dep"]["values"] = [
+                dep if dep is not None else MISSING_DEP_ for dep in token_annotations["dep"]["values"]
+            ]
+        if token_annotations["dep"]["values"] and not token_annotations["head"]["values"]:
+            token_annotations["head"]["values"] = [0] * len(token_annotations["dep"]["values"])
+        if token_annotations["head"]["values"] and not token_annotations["dep"]["values"]:
+            raise ValueError(Errors.E1017)
+        if token_annotations["pos"]["values"] is not None:
+            for pp in set(token_annotations["pos"]["values"]):
+                if pp not in parts_of_speech.IDS:
+                    raise ValueError(Errors.E1021.format(pp=pp))
+
+        headings = []
+        values = []
+        annotations = [token_annotations[key]["values"] for key in annotation_types]
+        possible_headings = [POS, HEAD, DEP, LEMMA, TAG, MORPH]
+        for a, annot in enumerate(annotations):
+            if annot is not None:
+                if len(annot) != len(words):
+                    raise ValueError(Errors.E189)
+                headings.append(possible_headings[a])
+                if annot is not token_annotations["head"]["values"]:
+                    values.extend(annot)
+        for value in values:
+            if value is not None:
+                self.vocab.strings.add(value)
```

Comment:
I think you can skip this if you switch to `self.vocab.strings.add(annot[i])` below?

Reply:
Let me know if the latest changes make sense to you.